### PR TITLE
feat(expr): implement `<expr> [ NOT ] SIMILAR TO <pat> [ ESCAPE <esc_text> ]`

### LINKED ISSUE #8966: Implement `[NOT] SIMILAR TO`
# Syntax

``` sql
string SIMILAR TO pattern [ESCAPE escape-character]
```

# Description

The `SIMILAR TO` predicate in Postgres is used to determine if a string
matches a specified pattern using SQL regular expressions. The pattern
can include various metacharacters, such as `|`, `*`, `+`, and others.
Optionally, you can specify an `ESCAPE` character to treat
metacharacters as literals within the pattern.

# Parameters

  - `string`: The input string to be compared against the pattern.

  - `pattern`: The SQL regular expression pattern to match against the
    input string.

  - `escape-character` \[optional\]: The character to be used as an
    escape character when you want to treat metacharacters as literals
    in the pattern.

# Returns

The predicate returns `true` if the input string matches the specified
pattern, and `false` otherwise.

# Examples

``` sql
SELECT 'abc' SIMILAR TO 'a|b|c'; -- Returns false
SELECT 'abc' SIMILAR TO '(a|b|c)+'; -- Returns true
SELECT 'a_b' SIMILAR TO 'a$_b' ESCAPE '$'; -- Returns true
```

# See Also

  - `LIKE`: Tests if a string matches a pattern using simple pattern
    matching with wildcards.

  - `ILIKE`: Case-insensitive version of the `LIKE` operator.

  - `POSIX Regular Expressions`: For more complex pattern matching, you
    can use POSIX regular expressions with operators like `~`, `!~`,
    `~*`, and `!~*`.

# Compatibility

The `SIMILAR TO` predicate is compatible with PostgreSQL and is part of
the SQL standard.

### PR DESCRIPTION
**Pull request by jetjinser**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

resolves #8966.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Support new clause: `<expr> [ NOT ] SIMILAR TO <pat> [ ESCAPE <esc_text> ]`.

</details>

### CODE REVIEW

**Review by wangrunji0408 — COMMENTED** [review]
Rest LGTM. Thanks!

File: src/meta/src/controller/rename.rs [hunk 1]
```diff
@@ -280,6 +280,10 @@ impl QueryRewriter<'_> {
                 self.visit_expr(low);
                 self.visit_expr(high);
             }
+            Expr::SimilarTo { expr, pat, .. } => {
+                self.visit_expr(expr);
+                self.visit_expr(pat);
+            }
```
[the remark this answers is not in the document]
Comment (jetjinser) [reply src/meta/src/controller/rename.rs:286]:
```rust
if let Some(e) = esc_text {
    self.visit_expr(e);
}
```
I don’t know much about meta node, is this okay?

File: src/sqlparser/src/parser.rs [hunk 2]
```diff
@@ -1685,6 +1687,23 @@ impl Parser {
         })
     }
 
+    /// Parses `SIMILAR TO <pat> [ ESCAPE <esc_text> ]`
+    pub fn parse_similar_to(&mut self, expr: Expr, negated: bool) -> Result<Expr, ParserError> {
+        let pat = self.parse_subexpr(Precedence::Between)?;
+        let esc_text = if self.parse_keyword(Keyword::ESCAPE) {
+            Some(Box::new(self.parse_subexpr(Precedence::Between)?))
```
[the remark this answers is not in the document]
Comment (jetjinser) [reply src/sqlparser/src/parser.rs:1694]: They have the same precedence, maybe the name `Precedence::Between` isn't quite suitable?

File: src/meta/src/controller/rename.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/meta/src/controller/rename.rs:286]: yes

**Review by wangrunji0408 — APPROVED** [review]
LGTM!

**Pull request merged**